File: typescript/src/shared/model-prices.ts
import { isPriceTable, type PriceTable, type PriceInput, type PriceParameter } from "./pricing.js";
import type { ChannelParameters } from "./parameter-rules.js";
import { getModelContract, getModelOperationContract } from "./model-contracts.js";
import { validateGeneration, type StandardImageQuoteRequest, type StandardVideoQuoteRequest } from "./standard.js";

export interface ModelPriceFilter {
  resolution?: string;
  reference_count?: number;
}

export interface ModelPrices {
  channel_parameters?: readonly ChannelParameters[];
  id: string;
  object: "model_prices";
  operation: "generate_image" | "generate_video";
  input_mode: "text" | "image" | "reference";
  contract_version: string;
  /** Coverage applies only to the requested filter, not the entire model. */
  coverage: "enumerated" | "partial";
  issues: string[];
  filter?: ModelPriceFilter;
  prices: {
    primary: PriceTable;
    max: PriceTable;
    expires_at: number;
    unavailable: {id: string; model: string; operation: string; conditions: Record<string, PriceParameter[]>; basis: "primary" | "max"; reason: string}[];
  };
}

export function modelPricesPath(model: string, operation: ModelPrices["operation"], inputMode: ModelPrices["input_mode"], filter: ModelPriceFilter = {}): string {
  if (typeof model !== "string") throw new Error("model_prices_request_invalid");
  const id = getModelContract(model)?.id ?? model;
  if (!/^[a-z0-9](?:[a-z0-9_.-]*[a-z0-9])?\/[a-z0-9](?:[a-z0-9_.-]*[a-z0-9])?$/.test(id) || id.split("/").some(x => x.length > 100) ||
      !["generate_image", "generate_video"].includes(operation) || !["text", "image", "reference"].includes(inputMode)) throw new Error("model_prices_request_invalid");
  if (!validFilter(filter, inputMode)) throw new Error("model_prices_request_invalid");
  return `/v1/models/${id}?view=pricing&operation=${operation}&input_mode=${inputMode}` +
    (filter.resolution === undefined ? "" : `&resolution=${encodeURIComponent(filter.resolution.toLowerCase())}`) +
    (filter.reference_count === undefined ? "" : `&reference_count=${filter.reference_count}`);
}

function validFilter(filter: ModelPriceFilter, mode: string): boolean {
  return !!filter && typeof filter === "object" && !Array.isArray(filter) && Object.keys(filter).every(k => k === "resolution" || k === "reference_count") &&
    (filter.resolution === undefined || (typeof filter.resolution === "string" && /^[a-z0-9.]{1,32}$/i.test(filter.resolution))) &&
    (filter.reference_count === undefined || (mode === "image" && Number.isInteger(filter.reference_count) && filter.reference_count >= 1 && filter.reference_count <= 512));
}

function matchesFilter(conditions: Record<string, PriceParameter[]>, filter: ModelPriceFilter): boolean {
  if (filter.resolution !== undefined && (conditions.resolution?.length !== 1 || conditions.resolution[0] !== filter.resolution.toLowerCase())) return false;
  if (filter.reference_count !== undefined) {
    let count = 0;
    for (const role of ["source_image", "first_frame", "last_frame", "reference_image", "reference_video", "reference_audio"]) {
      const values = conditions[`${role}_count`];
      if (values?.length !== 1 || typeof values[0] !== "number" || !Number.isSafeInteger(values[0]) || values[0] < 0 || values[0] > 512) return false;
      count += values[0];
    }
    if (count !== filter.reference_count) return false;
  }
  return true;
}

export function validateModelPrices(value: unknown, model: string, operation: ModelPrices["operation"], inputMode: ModelPrices["input_mode"], filter: ModelPriceFilter = {}): ModelPrices {
  const fail = (): never => { throw new Error("model_prices_response_invalid"); };
  const v = value as ModelPrices;
  if (!validFilter(filter,inputMode) || (v?.filter !== undefined && !validFilter(v.filter,inputMode)) ||
      (v?.filter?.resolution ?? "") !== (filter.resolution?.toLowerCase() ?? "") || v?.filter?.reference_count !== filter.reference_count) return fail();
  if (!v || v.object !== "model_prices" || v.id !== (getModelContract(model)?.id ?? model) || v.operation !== operation || v.input_mode !== inputMode ||
      typeof v.contract_version !== "string" || !v.contract_version || !["enumerated", "partial"].includes(v.coverage) || !Array.isArray(v.issues) ||
      !v.issues.every(x => typeof x === "string") || !v.prices || !Number.isSafeInteger(v.prices.expires_at) || v.prices.expires_at <= 0 || !Array.isArray(v.prices.unavailable)) return fail();
  for (const table of [v.prices.primary, v.prices.max]) {
    if (!isPriceTable(table) || table.purpose !== "yir-cost" || table.unit !== "USD" || table.scale !== 1_000_000 ||
        table.rows.some(row => row.model !== v.id || row.operation !== operation || row.conditions.input_mode?.length !== 1 || row.conditions.input_mode[0] !== inputMode || !matchesFilter(row.conditions,filter))) return fail();
  }
  for (const row of v.prices.unavailable) {
    if (!row || !["primary", "max"].includes(row.basis) || typeof row.reason !== "string" || !row.reason || row.model !== v.id || row.operation !== operation ||
        !isPriceTable({...v.prices.primary, rows: [{...row, kind: "exact", price: {type: "total", amount: "0"}}]}) ||
        row.conditions.input_mode?.length !== 1 || row.conditions.input_mode[0] !== inputMode || !matchesFilter(row.conditions,filter)) return fail();
  }
  return v;
}

/** Map a Standard request to the server price table's dimensions without I/O.
 * Uses bundled defaults; table matching still fails closed for unsupported combinations.
 * Does not modify the request or convert procurement cost to retail credits.
 */
export function buildPriceInput(operation: ModelPrices["operation"], request: StandardImageQuoteRequest | StandardVideoQuoteRequest, version: string): PriceInput {
  validateGeneration(operation, request);
  if (request.routing !== undefined) throw new Error("model_price_routing_override_unsupported");
  const contract = getModelOperationContract(request.model, operation, request.input.type)!;
  const parameters: Record<string, PriceParameter> = {};
  for (const rule of contract.parameters) {
    if (rule.name === "seed") continue;
    const value = request.parameters[rule.name] ?? rule.default;
    if (value !== undefined) parameters[rule.name] = value as PriceParameter;
  }
  parameters.input_mode = request.input.type;
  parameters.resolution = String(parameters.resolution).toLowerCase();
  if (operation === "generate_image") parameters.quality ??= "";
  else parameters.duration ??= 0;
  for (const role of ["source_image", "first_frame", "last_frame", "reference_image", "reference_video", "reference_audio"]) parameters[`${role}_count`] = 0;
  if (request.input.type !== "text") for (const reference of request.input.references) {
    const key = `${reference.role}_count`;
    parameters[key] = (parameters[key] as number) + 1;
  }
  return {model: getModelContract(request.model)!.id, operation, parameters, version};
}

File: go/model_prices.go
package yir

import (
	"context"
	"errors"
	"net/http"
	"regexp"
	"strconv"
	"strings"
)

type ModelPriceFilter struct {
	Resolution     string `json:"resolution,omitempty"`
	ReferenceCount *int   `json:"reference_count,omitempty"`
}

var modelPriceResolutionPattern = regexp.MustCompile(`^[a-zA-Z0-9.]{1,32}$`)

func (f ModelPriceFilter) valid(mode string) bool {
	return (f.Resolution == "" || modelPriceResolutionPattern.MatchString(f.Resolution)) && (f.ReferenceCount == nil || (mode == "image" && *f.ReferenceCount >= 1 && *f.ReferenceCount <= 512))
}

func (f ModelPriceFilter) matches(conditions map[string][]any) bool {
	if f.Resolution != "" && (len(conditions["resolution"]) != 1 || conditions["resolution"][0] != strings.ToLower(f.Resolution)) {
		return false
	}
	if f.ReferenceCount != nil {
		count := int64(0)
		for _, role := range []string{"source_image", "first_frame", "last_frame", "reference_image", "reference_video", "reference_audio"} {
			values := conditions[role+"_count"]
			if len(values) != 1 {
				return false
			}
			n, ok := priceNumber(values[0])
			if !ok || n < 0 || n > 512 {
				return false
			}
			count += n
		}
		if count != int64(*f.ReferenceCount) {
			return false
		}
	}
	return true
}

type ModelPrices struct {
	ChannelParameters []ChannelParameters `json:"channel_parameters,omitempty"`
	ID                string              `json:"id"`
	Object            string              `json:"object"`
	Operation         string              `json:"operation"`
	InputMode         string              `json:"input_mode"`
	ContractVersion   string              `json:"contract_version"`
	Coverage          string              `json:"coverage"`
	Issues            []string            `json:"issues"`
	Filter            *ModelPriceFilter   `json:"filter,omitempty"`
	Prices            ModelPriceTables    `json:"prices"`
}

type ModelPriceTables struct {
	Primary     PriceTable              `json:"primary"`
	Max         PriceTable              `json:"max"`
	ExpiresAt   int64                   `json:"expires_at"`
	Unavailable []ModelPriceUnavailable `json:"unavailable"`
}

type ModelPriceUnavailable struct {
	ID         string           `json:"id"`
	Model      string           `json:"model"`
	Operation  string           `json:"operation"`
	Conditions map[string][]any `json:"conditions"`
	Basis      string           `json:"basis"`
	Reason     string           `json:"reason"`
}

var modelPriceSegmentPattern = regexp.MustCompile(`^[a-z0-9](?:[a-z0-9_.-]*[a-z0-9])?$`)

// GetModelPrices makes one scoped read. Applications own refresh, policy-isolated
// caching and retail conversion; request overrides require request-level Quote.
func (c *Client) GetModelPrices(ctx context.Context, model, operation, inputMode string, filters ...ModelPriceFilter) (ModelPrices, error) {
	if len(filters) > 1 {
		return ModelPrices{}, errors.New("model_prices_request_invalid")
	}
	filter := ModelPriceFilter{}
	if len(filters) == 1 {
		filter = filters[0]
	}
	if !filter.valid(inputMode) {
		return ModelPrices{}, errors.New("model_prices_request_invalid")
	}
	filter.Resolution = strings.ToLower(filter.Resolution)
	query := ""
	if filter.Resolution != "" {
		query += "&resolution=" + filter.Resolution
	}
	if filter.ReferenceCount != nil {
		query += "&reference_count=" + strconv.Itoa(*filter.ReferenceCount)
	}
	if contract, ok := GetModelContract(model); ok {
		model = contract.ID
	}
	parts := strings.Split(model, "/")
	if len(parts) != 2 || len(parts[0]) > 100 || len(parts[1]) > 100 || !modelPriceSegmentPattern.MatchString(parts[0]) || !modelPriceSegmentPattern.MatchString(parts[1]) ||
		(operation != "generate_image" && operation != "generate_video") || (inputMode != "text" && inputMode != "image" && inputMode != "reference") {
		return ModelPrices{}, errors.New("model_prices_request_invalid")
	}
	var result ModelPrices
	err := c.do(ctx, http.MethodGet, "/v1/models/"+model+"?view=pricing&operation="+operation+"&input_mode="+inputMode+query, "", nil, &result)
	if err == nil {
		err = result.Validate(model, operation, inputMode, filter)
	}
	if err != nil {
		return ModelPrices{}, err
	}
	return result, nil
}

func (v ModelPrices) Validate(model, operation, inputMode string, filters ...ModelPriceFilter) error {
	invalid := errors.New("model_prices_response_invalid")
	if len(filters) > 1 {
		return invalid
	}
	want, got := ModelPriceFilter{}, ModelPriceFilter{}
	if len(filters) == 1 {
		want = filters[0]
	}
	if v.Filter != nil {
		got = *v.Filter
	}
	if !want.valid(inputMode) || !got.valid(inputMode) || got.Resolution != strings.ToLower(want.Resolution) || (got.ReferenceCount == nil) != (want.ReferenceCount == nil) {
		return invalid
	}
	if got.ReferenceCount != nil && *got.ReferenceCount != *want.ReferenceCount {
		return invalid
	}
	if v.Object != "model_prices" || v.ID != model || v.Operation != operation || v.InputMode != inputMode || v.ContractVersion == "" ||
		(v.Coverage != "enumerated" && v.Coverage != "partial") || v.Issues == nil || v.Prices.Unavailable == nil || v.Prices.ExpiresAt <= 0 || v.Prices.ExpiresAt > maxPriceParameter {
		return invalid
	}
	for _, table := range []PriceTable{v.Prices.Primary, v.Prices.Max} {
		if !table.Valid() || table.Purpose != "yir-cost" || table.Unit != "USD" || table.Scale != 1000000 {
			return invalid
		}
		for _, row := range table.Rows {
			if row.Model != model || row.Operation != operation || len(row.Conditions["input_mode"]) != 1 || row.Conditions["input_mode"][0] != inputMode || !want.matches(row.Conditions) {
				return invalid
			}
		}
	}
	for _, row := range v.Prices.Unavailable {
		if (row.Basis != "primary" && row.Basis != "max") || row.Reason == "" || row.Model != model || row.Operation != operation || len(row.Conditions["input_mode"]) != 1 || row.Conditions["input_mode"][0] != inputMode {
			return invalid
		}
		table := v.Prices.Primary
		table.Rows = []PriceRow{{ID: row.ID, Model: row.Model, Operation: row.Operation, Conditions: row.Conditions, Kind: "exact", Price: TablePrice{Type: "total", Amount: "0"}}}
		if !table.Valid() || !want.matches(row.Conditions) {
			return invalid
		}
	}
	return nil
}

// BuildPriceInput applies bundled defaults and maps a Standard request to local
// table conditions. It never changes the original request or performs I/O.
func BuildPriceInput(operation string, request GenerationRequest, version string) (PriceInput, error) {
	if err := ValidateGeneration(operation, request); err != nil {
		return PriceInput{}, err
	}
	if request.Routing != nil {
		return PriceInput{}, errors.New("model_price_routing_override_unsupported")
	}
	contract, _ := GetModelOperationContract(request.Model, operation, request.Input.Type)
	parameters := map[string]any{}
	for _, rule := range contract.Parameters {
		if rule.Name == "seed" {
			continue
		}
		value, ok := request.Parameters[rule.Name]
		if !ok {
			value = rule.Default
		}
		if value != nil {
			parameters[rule.Name] = value
		}
	}
	parameters["input_mode"] = request.Input.Type
	parameters["resolution"] = strings.ToLower(parameters["resolution"].(string))
	if operation == "generate_image" {
		if _, ok := parameters["quality"]; !ok {
			parameters["quality"] = ""
		}
	} else if _, ok := parameters["duration"]; !ok {
		parameters["duration"] = 0
	}
	for _, role := range []string{"source_image", "first_frame", "last_frame", "reference_image", "reference_video", "reference_audio"} {
		parameters[role+"_count"] = 0
	}
	for _, ref := range request.Input.References {
		key := ref.Role + "_count"
		parameters[key] = parameters[key].(int) + 1
	}
	model, _ := GetModelContract(request.Model)
	return PriceInput{Model: model.ID, Operation: operation, Parameters: parameters, Version: version}, nil
}
